src/Lykke.Job.CandlesHistoryWriter.Services/HistoryMigration/TradesMigrationManager.cs

```diff
@@ -0,0 +1,139 @@
+            )
+        {
+            _candlesHistoryRepository = candlesHistoryRepository;
+            _log = log;
+
+            _sqlConnString = sqlConnString;
+            _sqlQueryBatchSize = sqlQueryBatchSize;
+
+            Health = null;
+        }
+
+        public void Migrate(ITradesMigrationRequest request)
+        {
+            Task.Run(() => DoMigrateAsync(request).Wait());
+        }
+
+        private async Task DoMigrateAsync(ITradesMigrationRequest request)
+        {
+            Health = new TradesMigrationHealthReport(_sqlQueryBatchSize);
+
+            foreach (var migrationItem in request.MigrationItems)
+            {
+                Health.AssetReportItems[migrationItem.AssetId] = new TradesMigrationHealthReportItem(migrationItem.OffsetFromTop);
+
+                using (var sqlRepo = new TradesSqlHistoryRepository(_sqlConnString, _sqlQueryBatchSize, _log,
+                    migrationItem.OffsetFromTop, migrationItem.AssetId))
+                {
+                    await _log.WriteInfoAsync(nameof(TradesMigrationManager), nameof(Migrate),
+                        $"Starting migration for {migrationItem.AssetId}, except firts {migrationItem.OffsetFromTop} records.");
+
+                    while (true)
+                    {
+                        var tradesBatch = await sqlRepo.GetNextBatchAsync();
+                        var batchCount = tradesBatch.Count();
+
+                        if (batchCount == 0) break;
+
+                        TradesCandleBatch smallerCandles = null;
+
+                        var batchInsertTasks = new List<Task>();
+
+                        // Now we can build sequentally candle batches for all the stored time intervals.
+                        // To avoid looking through the whole trades batch several times, we derive the
+                        // bigger time intervals from the smaller ones. For example, let's assume that we
+                        // generate 1000 sec candles from our batch of 2381 trades. Then, for calculating
+                        // minute candles we need to iterate only 1000 second candles, and this will give
+                        // us 16 minute candles. Next, we iterate 16 minute candles to get 1 hour candle,
+                        // and so on, instead of looking through 2381 trades for each time interval.
+                        foreach (var interval in Constants.StoredIntervals)
+                        {
+                            // It's important for Constants.StoredIntervals to be ordered by time period increase,
+                            // because we will calculate candles for each period based on previous period candles.
+                            var currentCandles = interval == CandleTimeInterval.Sec
+                                ? new TradesCandleBatch(migrationItem.AssetId, interval, tradesBatch)
+                                : new TradesCandleBatch(migrationItem.AssetId, interval, smallerCandles);
```

Comment:
Since you have multiple TradesCandleBatch ctors with different meaning, will be good to make ctors private and expose public static methods with meaningfull (from the point of domain specific language) names, for example MakeFromTrades and DeriveFromSmallerInterval. Simple new tells nothing to code reader about how this one new differs from another one, while TradesCandleBatch.MakeFromTrades() and TradesCandleBatch.DeriveFromSmallerInterval() tell what exactly you want to do.